Index: common.h
===================================================================
RCS file: /cvsroot/pkgsrc/pkgtools/cwrappers/files/bin/common.h,v
retrieving revision 1.9
diff -u -p -u -r1.9 common.h
--- common.h	7 Nov 2021 12:38:12 -0000	1.9
+++ common.h	7 Nov 2021 12:39:14 -0000
@@ -121,6 +121,6 @@ void	fixup_libtool(struct arglist *);
 void	mi_vector_hash(const void *, size_t, uint32_t, uint32_t[3]);
 #endif
 
-#define isabs(a) ((a) == '/')
+#define isabs(a) ((a) == '/' || (a) == '$')
 
 #endif
Index: generic-transform-cc.c
===================================================================
RCS file: /cvsroot/pkgsrc/pkgtools/cwrappers/files/bin/generic-transform-cc.c,v
retrieving revision 1.5
diff -u -p -u -r1.5 generic-transform-cc.c
--- generic-transform-cc.c	7 Nov 2021 12:38:12 -0000	1.5
+++ generic-transform-cc.c	7 Nov 2021 12:39:14 -0000
@@ -462,6 +462,8 @@ generic_transform_cc(struct arglist *arg
 				if (len >= 6 &&
 				    strcmp(path + len - 6, "/.libs") == 0)
 					continue;
+				if (strncmp(path, "$ORIGIN", 7) == 0)
+					continue;
 			}
 			if (opt_arg)
 				argument_unlink(args, &opt_arg);
