Index: java-vm.mk
===================================================================
RCS file: /cvsroot/pkgsrc/mk/java-vm.mk,v
retrieving revision 1.14
diff -u -r1.14 java-vm.mk
--- java-vm.mk	2003/05/25 18:39:12	1.14
+++ java-vm.mk	2003/07/06 20:27:15
@@ -43,9 +43,9 @@
 # Note: The wonka configuration is still under development
 #
 .if defined(USE_JAVA2) && !empty(USE_JAVA2:M[yY][eE][sS])
-_PKG_JVMS?=		sun-jdk13 sun-jdk14 blackdown-jdk13 wonka
+_PKG_JVMS?=		sun-jdk13 sun-jdk14 blackdown-jdk13 wonka jdk12 jdk13
 .else
-_PKG_JVMS?=		jdk sun-jdk13 sun-jdk14 blackdown-jdk13 kaffe wonka
+_PKG_JVMS?=		jdk sun-jdk13 sun-jdk14 blackdown-jdk13 kaffe wonka jdk12 jdk13
 .endif
 
 # To be deprecated: if PKG_JVM is explicitly set, then use it as the
@@ -78,6 +78,10 @@
 # These lists are copied from the JVM package Makefiles.
 _ONLY_FOR_PLATFORMS.jdk= \
 	NetBSD-*-i386 Linux-*-i[3-6]86
+_ONLY_FOR_PLATFORMS.jdk12= \
+	NetBSD-1.[6-9]*-i386 NetBSD-2*-i386
+_ONLY_FOR_PLATFORMS.jdk13= \
+	NetBSD-1.[6-9]*-i386 NetBSD-2*-i386
 _ONLY_FOR_PLATFORMS.blackdown-jdk13= \
 	NetBSD-*-i386 NetBSD-*-powerpc NetBSD-*-sparc \
 	Linux-*-i[3-6]86 Linux-*-powerpc Linux-*-sparc
@@ -104,6 +108,8 @@
 .endfor
 
 _JAVA_PKGBASE.jdk=		jdk
+_JAVA_PKGBASE.jdk12=		jdk12
+_JAVA_PKGBASE.jdk13=		jdk13
 _JAVA_PKGBASE.sun-jdk13=	sun-jdk13
 _JAVA_PKGBASE.sun-jdk14=	sun-jdk14
 _JAVA_PKGBASE.blackdown-jdk13=	blackdown-jdk13
@@ -180,6 +186,8 @@
 .endif
 
 BUILDLINK_DEPENDS.jdk?=			jdk-[0-9]*
+BUILDLINK_DEPENDS.jdk12?=		jdk12-[0-9]*
+BUILDLINK_DEPENDS.jdk13?=		jdk13-[0-9]*
 BUILDLINK_DEPENDS.sun-jdk13?=		sun-jdk13-[0-9]*
 BUILDLINK_DEPENDS.sun-jre13?=		sun-jre13-[0-9]*
 BUILDLINK_DEPENDS.sun-jdk14?=		sun-jdk14-[0-9]*
@@ -190,6 +198,8 @@
 BUILDLINK_DEPENDS.wonka?=		wonka-[0-9]*
 
 _JRE.jdk=		jdk
+_JRE.jdk12=		jdk12
+_JRE.jdk13=		jdk13
 _JRE.sun-jdk13=		sun-jre13
 _JRE.sun-jdk14=		sun-jre14
 _JRE.blackdown-jdk13=	blackdown-jre13
@@ -202,6 +212,14 @@
 _JDK_PKGSRCDIR=		../../lang/jdk
 _JRE_PKGSRCDIR=		${_JDK_PKGSRCDIR}
 _JAVA_HOME_DEFAULT=	${LOCALBASE}/java/jdk-1.1.8
+.elif ${_PKG_JVM} == "jdk12"
+_JDK_PKGSRCDIR=		../../wip/jdk12
+_JRE_PKGSRCDIR=		${_JDK_PKGSRCDIR}
+_JAVA_HOME_DEFAULT=	${LOCALBASE}/java/jdk-1.2.2
+.elif ${_PKG_JVM} == "jdk13"
+_JDK_PKGSRCDIR=		../../wip/jdk13
+_JRE_PKGSRCDIR=		${_JDK_PKGSRCDIR}
+_JAVA_HOME_DEFAULT=	${LOCALBASE}/java/jdk-1.3.1
 .elif ${_PKG_JVM} == "sun-jdk13"
 _JDK_PKGSRCDIR=		../../lang/sun-jdk13
 _JRE_PKGSRCDIR=		../../lang/sun-jre13
