$NetBSD: egg-anthy.readme.diff,v 1.1.1.1 2010/08/04 12:14:16 makoto Exp $

--- /usr/pkg/share/anthy/egg-anthy.readme	2010-08-03 16:50:45.000000000 +0900
+++ egg-anthy/egg-anthy.readme	2010-08-03 21:17:03.000000000 +0900
@@ -50,7 +50,7 @@
  egg-anthy 付きのバージョンか否かで、方法が変わります。
 
 3.2.1 "C-uC-\" で "japanese-egg-anthy" が出てこない人
- 2.1 を参考にして、一から Tamago をインストールしてください。
+ 3.1 を参考にして、一から Tamago をインストールしてください。
 
 3.2.2 "C-uC-\" で "japanese-egg-anthy" が出てくる人
  (環境によって違うと思いますが)/usr/local/local/share/emacs/site-lisp/egg/egg
@@ -102,7 +102,7 @@
 (2) anthy-6131 付属の anthy-dic-tool がサポートしている品詞はすべて egg か
     ら登録することが出来ます。
 
-  動詞の登録は癖があります。例えば「焼き込む」を登録するときは、漢字:『書
+  動詞の登録は癖があります。例えば「焼き込む」を登録するときは、漢字:『焼
   き込』、読み:『やきこ』、活用:『マ行五段活用』の様に活用語尾を外して登録
   します。
 
@@ -141,7 +141,7 @@
 6. UTF-8 対応
  最近の Anthy では、Anthy 内部が UNICODE 化し、安定版では anthy-8700 以降
  の anthy-agent, anthy-dic-tool でも "--utf8" というオプションを使えば
- UTF-8 で通信することができます。お使いの Emacs が Emacs-22 または
+ UTF-8 で通信することができます。お使いの Emacs が Emacs-22 以降または
  Emacs-21 + Mule-UCS のときは、Anthy と UTF-8 で通信することができます。
 
 6.1 設定
