# $NetBSD$
#
###########################################################
#                  Generated by fbsd2pkg                  #
#              Thu Sep 13 14:20:25 CDT 2018               #
###########################################################

DISTNAME=	salmon-${PV}
DIST_SUBDIR=	salmon
CATEGORIES=	biology
MASTER_SITES=	${MASTER_SITE_GITHUB:=COMBINE-lab/} \
		https://github.com/COMBINE-lab/RapMap/archive/ \
		https://github.com/COMBINE-lab/libgff/archive/ \
		https://github.com/COMBINE-lab/staden-io_lib/archive/
DISTFILES=	salmon-0.11.3.tar.gz salmon-v0.11.3.zip v1.1.tar.gz v1.14.8.1.tar.gz
EXTRACT_ONLY=	salmon-0.11.3.tar.gz v1.14.8.1.tar.gz
GITHUB_PROJECT=	salmon
GITHUB_TAG=	v${PV}

MAINTAINER=	bacon@NetBSD.org
HOMEPAGE=	https://github.com/COMBINE-lab/salmon
COMMENT=	Transcript-level quantification of RNA-seq from lightweight alignments
LICENSE=	gnu-gpl-v3

SUBST_CLASSES+=		tbb
SUBST_STAGE.tbb=	post-patch
SUBST_SED.tbb+=		-e 's|tbbmalloc_proxy||g'
SUBST_FILES.tbb+=	${WRKSRC}/CMakeLists.txt

SUBST_CLASSES+=		staden
SUBST_STAGE.staden=	post-patch
SUBST_SED.staden+=	-e 's|&& make|&& gmake|g'
SUBST_FILES.staden+=	${WRKSRC}/CMakeFiles/libstadenio.dir/build.make

# Test and change if necessary.
# MAKE_JOBS_SAFE=	no

DEPENDS=	bash>=0:../../shells/bash \
		autoconf>=0:../../devel/autoconf \
		unzip>=0:../../archivers/unzip

USE_LANGUAGES=	c c++
USE_TOOLS+=	cmake gmake
REPLACE_BASH=	scripts/*.sh
REPLACE_PYTHON=	scripts/*.py
USE_CMAKE=	yes
CMAKE_CONFIGURE_ARGS+=	-DDISTDIR:PATH=${DISTDIR} -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON

CONFIGURE_ENV+=	DISTDIR=${DISTDIR} DIST_SUBDIR=salmon
LDFLAGS+=	-lz

PV=		0.11.3

.include "../../devel/libgetopt/buildlink3.mk"
.include "../../devel/zlib/buildlink3.mk"
.include "../../devel/boost-libs/buildlink3.mk"
.include "../../wip/cereal/buildlink3.mk"
.include "../../wip/tbb/buildlink3.mk"
.include "../../mk/bsd.pkg.mk"
