Index: PLIST
===================================================================
RCS file: /cvsroot/pkgsrc/graphics/cairo/PLIST,v
retrieving revision 1.17
diff -u -b -r1.17 PLIST
--- PLIST	15 Jun 2007 17:47:59 -0000	1.17
+++ PLIST	25 Mar 2008 20:46:55 -0000
@@ -8,7 +8,9 @@
 include/cairo/cairo-ps.h
 ${CARBON}include/cairo/cairo-quartz.h
 include/cairo/cairo-svg.h
+${XCB}include/cairo/cairo-xcb-xrender.h
 ${X11}include/cairo/cairo-xlib-xrender.h
+${XCB}include/cairo/cairo-xcb.h
 ${X11}include/cairo/cairo-xlib.h
 include/cairo/cairo.h
 lib/libcairo.la
@@ -20,6 +22,7 @@
 lib/pkgconfig/cairo-ps.pc
 ${CARBON}lib/pkgconfig/cairo-quartz.pc
 lib/pkgconfig/cairo-svg.pc
+${XCB}lib/pkgconfig/cairo-xcb.pc
 ${X11}lib/pkgconfig/cairo-xlib-xrender.pc
 ${X11}lib/pkgconfig/cairo-xlib.pc
 lib/pkgconfig/cairo.pc
Index: Makefile
===================================================================
RCS file: /cvsroot/pkgsrc/graphics/cairo/Makefile,v
retrieving revision 1.59
diff -u -b -r1.59 Makefile
--- Makefile	12 Mar 2008 20:48:11 -0000	1.59
+++ Makefile	25 Mar 2008 20:46:55 -0000
@@ -16,7 +16,7 @@
 USE_TOOLS+=		pkg-config
 PKGCONFIG_OVERRIDE+=	src/cairo.pc.in src/cairo-backend.pc.in
 GNU_CONFIGURE=		yes
-CONFIGURE_ARGS+=	--disable-xcb
+
 CONFIGURE_ARGS+=	--enable-glitz
 CONFIGURE_ARGS+=	--enable-pdf
 CONFIGURE_ARGS+=	--enable-ps
@@ -38,6 +38,17 @@
 
 TEST_TARGET=			check
 
+.include "../../x11/libX11/buildlink3.mk"
+
+.if !empty(PKG_BUILD_OPTIONS.libX11:U:Mxcb)
+CONFIGURE_ARGS+=	--enable-xcb
+PLIST_SUBST+=		XCB=""
+.include "../../wip/xcb-util/buildlink3.mk"
+.else
+CONFIGURE_ARGS+=	--disable-xcb
+PLIST_SUBST+=		XCB="@comment "
+.endif
+
 .include "../../devel/zlib/buildlink3.mk"
 .include "../../fonts/fontconfig/buildlink3.mk"
 .include "../../graphics/freetype2/buildlink3.mk"
