diff -ur embedded_innodb-1.0.6.6750.orig/include/univ.i embedded_innodb-1.0.6.6750/include/univ.i
--- embedded_innodb-1.0.6.6750.orig/include/univ.i	2010-02-22 12:57:13.000000000 +0000
+++ embedded_innodb-1.0.6.6750/include/univ.i
@@ -185,7 +185,7 @@ by one. */
 
 /* Linkage specifier for non-static InnoDB symbols (variables and functions)
 that are only referenced from within InnoDB. */
-#if defined(__GNUC__) && (__GNUC__ >= 4) || defined(__INTEL_COMPILER)
+#if defined(__GNUC__) && (__GNUC__ >= 4) && !defined(sun) || defined(__INTEL_COMPILER)
 # define UNIV_INTERN __attribute__((visibility ("hidden")))
 #else
 # define UNIV_INTERN
