--- pyversion.mk.orig	2003-07-31 15:20:10.000000000 +0200
+++ pyversion.mk	2003-07-31 15:44:52.000000000 +0200
@@ -5,11 +5,11 @@
 
 .include "../../mk/bsd.prefs.mk"
 
-PYTHON_VERSION_DEFAULT?=	22
+PYTHON_VERSION_DEFAULT?=	23
 .if ${OPSYS} == "Darwin"
-PYTHON_VERSIONS_ACCEPTED=	22	# the only one that builds
+PYTHON_VERSIONS_ACCEPTED=	23 22	# the only one that builds
 .else
-PYTHON_VERSIONS_ACCEPTED?=	22 22pth 21 21pth 20
+PYTHON_VERSIONS_ACCEPTED?=	23 22 22pth 21 21pth 20
 .endif
 
 BUILDLINK_DEPENDS.python15?=		python15>=1.5
@@ -18,6 +18,7 @@
 BUILDLINK_DEPENDS.python21-pth?=	python21-pth>=2.1
 BUILDLINK_DEPENDS.python22?=		python22>=2.2
 BUILDLINK_DEPENDS.python22-pth?=	python22-pth>=2.2
+BUILDLINK_DEPENDS.python23?=		python23>=2.3
 
 # transform the list into individual variables
 .for pv in ${PYTHON_VERSIONS_ACCEPTED}
@@ -25,6 +26,9 @@
 .endfor
 
 # check what is installed
+.if exists(${LOCALBASE}/bin/python2.3)
+_PYTHON_VERSION_23_INSTALLED=	yes
+.endif
 .if exists(${LOCALBASE}/bin/python2.2)
 _PYTHON_VERSION_22_INSTALLED=	yes
 .endif
@@ -89,7 +93,13 @@
 #  PYPKGPREFIX: prefix to use in PKGNAME for extensions which can install
 #               to multiple Python versions
 #
-.if ${_PYTHON_VERSION} == "22"
+.if ${_PYTHON_VERSION} == "23"
+PYPKGSRCDIR=	../../wip/python23
+PYDEPENDENCY=	${BUILDLINK_DEPENDS.python22}:${PYPKGSRCDIR}
+PYPACKAGE=	python23
+PYVERSSUFFIX=	2.3
+PYPKGPREFIX=	py23
+.elif ${_PYTHON_VERSION} == "22"
 PYPKGSRCDIR=	../../lang/python22
 PYDEPENDENCY=	${BUILDLINK_DEPENDS.python22}:${PYPKGSRCDIR}
 PYPACKAGE=	python22
