--- srcdist.mk.orig	2003-07-31 19:28:56.000000000 +0200
+++ srcdist.mk	2003-07-31 19:30:28.000000000 +0200
@@ -2,7 +2,17 @@
 
 .include "../../lang/python/pyversion.mk"
 
-.if ${_PYTHON_VERSION} == "22" || ${_PYTHON_VERSION} == "22pth"
+.if ${_PYTHON_VERSION} == "23"
+
+DISTNAME=	Python-2.3
+EXTRACT_SUFX=	.tgz
+DISTINFO_FILE=	${.CURDIR}/../../wip/python23/distinfo
+PATCHDIR=	${.CURDIR}/../../wip/python23/patches
+PYSUBDIR=	Python-2.3
+WRKSRC=		${WRKDIR}/${PYSUBDIR}
+MASTER_SITES=	ftp://ftp.python.org/pub/python/2.3/
+
+.elif ${_PYTHON_VERSION} == "22" || ${_PYTHON_VERSION} == "22pth"
 
 DISTNAME=	Python-2.2.3
 EXTRACT_SUFX=	.tgz
